### PR TITLE
Incentivize based on contribution

### PR DESCRIPTION
**Pull request by 0xcodercrane**
Resolves https://github.com/ubiquity/ubiquibot/issues/272

### CODE REVIEW

File: src/utils/helpers.ts [hunk 1]
```diff
@@ -51,6 +51,16 @@ export const getAnalyticsMode = (parsedRepo: WideRepoConfig | undefined, parsedO
   }
 };
 
+export const getIncentiveMode = (parsedRepo: WideRepoConfig | undefined, parsedOrg: WideOrgConfig | undefined): boolean => {
```
Comment (pavlovcik) [inline src/utils/helpers.ts:54]:
```suggestion
export const getIncentiveMode = (parsedRepo?: WideRepoConfig, parsedOrg?: WideOrgConfig): boolean => {
```

File: src/handlers/processors.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ import { nullHandler } from "./shared";
 import { handleComment, issueClosedCallback } from "./comment";
 import { checkPullRequests } from "./assign/auto";
 import { createDevPoolPR } from "./pull-request";
+import { incentiveContribution } from "./payout";
```
Comment (pavlovcik) [inline src/handlers/processors.ts:9]: We should rename this function to `incentivizeComments` to be more precise with its purpose?

File: src/helpers/comment.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  Link: "link",
+  Text: "text",
+  Code: "code",
+} as const;
+export type MarkdownItem = (typeof MarkdownItem)[keyof typeof MarkdownItem];
+
+type MdastNode = {
+  type: string;
+  value: string;
+  children: MdastNode[];
```
Comment (pavlovcik) [inline src/helpers/comment.ts:17]: I'm surprised this recursive type definition works.

File: src/utils/private.ts [hunk 4]
```diff
@@ -51,17 +61,6 @@ export interface WideOrgConfig extends WideConfig {
   "private-key-encrypted"?: string;
 }
 
-export interface DataConfig {
```
Comment (rndquu) [inline src/utils/private.ts:54]: why was this type removed?

File: src/helpers/comment.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+export const MarkdownItem = {
+  Paragraph: "paragraph",
+  List: "list",
+  Link: "link",
+  Text: "text",
+  Code: "code",
+} as const;
+
+const PriceItem: Record<string, number> = {
```
Comment (rndquu) [inline src/helpers/comment.ts:13]: It should be moved to the bot's config. Should we create a separate issue for "add incentive config"?

File: src/helpers/comment.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  [MarkdownItem.Text]: 0.1,
+  [MarkdownItem.Link]: 0.5,
```
Comment (rndquu) [inline src/helpers/comment.ts:15]: Is there a difference between bullet and numbered lists in `mdast`? Or all of them are parsed as a `List` type? I mean is there smth like `List.Bullet` or `List.Numbered` ?

File: src/helpers/comment.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        result[key] = parsedContent[key];
+      }
+    }
+  }
+
+  let sum = 0;
+  for (const key of Object.keys(result)) {
+    const rewardValue = PriceItem[key];
+    const value = result[key];
+    if (key == MarkdownItem.Text) {
```
Comment (rndquu) [inline src/helpers/comment.ts:65]: Isn't the `Paragraph` type the same as the `Text` one? I mean can there possibly exist a paragraph node with a `value` field:
```
{
          "type": "paragraph",
          "value": "With the AI feature specified in the road map for ",
}
```

Or all of the paragraph nodes have children with `text` node types?

File: src/handlers/payout/post.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  const tokenSymbol = await getTokenSymbol(paymentToken, rpc);
+
+  // The mapping between gh handle and comment with a permit url
+  let reward: Record<string, string> = {};
+
+  // The mapping between gh handle and amount in ETH
+  let fallbackReward: Record<string, string> = {};
+  let comment: string = "";
+  for (const user of Object.keys(issueCommentsByUser)) {
```
Comment (rndquu) [inline src/handlers/payout/post.ts:42]: Bounty hunter and bot should not receive payments. Is it implemented?

File: src/helpers/comment.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      sum += value.length * rewardValue;
+    } else {
+      sum += rewardValue;
```
Comment (rndquu) [inline src/helpers/comment.ts:68]: Will contributor receive a payment for a quoted text?

File: src/helpers/comment.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
+import { fromMarkdown } from "mdast-util-from-markdown";
+import { gfmFromMarkdown } from "mdast-util-gfm";
+import { gfm } from "micromark-extension-gfm";
+
+export const MarkdownItem = {
```
Comment (rndquu) [inline src/helpers/comment.ts:5]: We should also add an image node type here

**Review by rndquu — COMMENTED** [review]
There are some questions in the comments but looks fine overall

@0xcodercrane pls fix the conflict

We should also create separate issues for:
1. Double the credit in the original post
2. Read repository BASE_RATE from the ubiquibot-config.yml and multiply the output

File: src/helpers/comment.ts (hunk 6 above)
Comment (pavlovcik) [reply src/helpers/comment.ts:15]: I think it is okay to consolidate both list types then.

File: src/handlers/payout/action.ts [hunk 11]
```diff
@@ -50,30 +55,24 @@ export const handleIssueClosed = async () => {
...
-    return comment;
-  }
-  return;
+  const payoutUrl = await generatePermit2Signature(recipient, priceInEth, issue.node_id);
+  const tokenSymbol = await getTokenSymbol(paymentToken, rpc);
+  const shortenRecipient = shortenEthAddress(recipient, `[ CLAIM ${priceInEth} ${tokenSymbol.toUpperCase()} ]`.length);
+  logger.info(`Posting a payout url to the issue, url: ${payoutUrl}`);
+  const comment = `### [ **[ CLAIM ${priceInEth} ${tokenSymbol.toUpperCase()} ]** ](${payoutUrl})\n` + "```" + shortenRecipient + "```";
+  await deleteLabel(issueDetailed.priceLabel!);
+  await addLabelToIssue("Paid");
```
Comment (pavlovcik) [inline src/handlers/payout/action.ts:76]:
```suggestion
  await addLabelToIssue("Permitted");
```

This needs to be permitted #301

File: src/helpers/comment.ts (hunk 5 above)
Comment (pavlovcik) [reply src/helpers/comment.ts:13]: Yes these should have configurable values. This variable should be named something like `CommentElementPricingDefaults`

**Review by pavlovcik — CHANGES_REQUESTED** [review]
You opened a pull request that doesn't include changes from a merged pull request a day prior. Either you're working off of outdated code when you started, or you have a problem in your development flow.

This is an issue if we're paying out bounty hunters to build features and then those features being excluded.

**Comment by 0xcodercrane** [comment]
1/ let me fix the conflicts by merging the devleoment branch.
2/ still doing refactors but will follow up the ones you guys commented

File: src/helpers/comment.ts (hunk 10 above)
Comment (0xcodercrane) [reply src/helpers/comment.ts:5]: https://github.com/ubiquity/ubiquibot/pull/305/commits/5a466cf96906ed0161567cdd520a7f2ccefd1087

File: src/helpers/comment.ts (hunk 5 above)
Comment (0xcodercrane) [reply src/helpers/comment.ts]: I guess that would be better to have a separate config. anyway pushed renaming

Comment (0xcodercrane) [reply src/helpers/comment.ts]: https://github.com/ubiquity/ubiquibot/pull/305/commits/a964f1fc7f02260fad10b7b577a0fd6090265199

File: src/helpers/comment.ts (hunk 6 above)
Comment (0xcodercrane) [reply src/helpers/comment.ts]: I couldn't find both `List.Bullet` and `List.Numbered` so I guess that's fine to consolidate both list types as of now.

File: src/handlers/payout/post.ts (hunk 8 above)
Comment (0xcodercrane) [reply src/handlers/payout/post.ts:42]: this was a nice catch! pushing fixes

File: src/handlers/payout/action.ts (hunk 11 above)
Comment (0xcodercrane) [reply src/handlers/payout/action.ts]: https://github.com/ubiquity/ubiquibot/pull/305/commits/cb91ad1344da8a3251b7adff323627272173d354

File: src/utils/helpers.ts [hunk 12]
```diff
@@ -41,6 +42,16 @@ export const getPriorityLabels = (parsedRepo: WideRepoConfig | undefined, parsed
   }
 };
 
+export const getCommentItemPrice = (parsedRepo: WideRepoConfig | undefined, parsedOrg: WideOrgConfig | undefined): CommentElementPricing => {
```
Comment (pavlovcik) [inline src/utils/helpers.ts:45]:
```suggestion
export const getCommentItemPrice = (parsedRepo?: WideRepoConfig, parsedOrg?: WideOrgConfig): CommentElementPricing => {
```

File: src/utils/helpers.ts [hunk 13]
```diff
@@ -41,6 +42,16 @@ export const getPriorityLabels = (parsedRepo: WideRepoConfig | undefined, parsed
...
+  if (parsedRepo && parsedRepo["comment-element-pricing"]) {
+    return parsedRepo["comment-element-pricing"];
+  } else if (parsedOrg && parsedOrg["comment-element-pricing"]) {
+    return parsedOrg["comment-element-pricing"];
+  } else {
+    return DefaultPriceConfig["commentElementPricing"];
+  }
+};
+
 export const getAutoPayMode = (parsedRepo: WideRepoConfig | undefined, parsedOrg: WideOrgConfig | undefined): boolean => {
```
Comment (pavlovcik) [inline src/utils/helpers.ts:55]:
```suggestion
export const getAutoPayMode = (parsedRepo?: WideRepoConfig, parsedOrg?: WideOrgConfig): boolean => {
```

File: src/handlers/payout/post.ts [hunk 14]
```diff
@@ -22,13 +22,21 @@ export const incentivizeComments = async () => {
   const payload = context.payload as Payload;
   const issue = payload.issue;
   if (!issue) {
-    logger.info(`Not an gh issue, skipping`);
+    logger.info(`Not an github issue, skipping`);
```
Comment (pavlovcik) [inline src/handlers/payout/post.ts:25]:
```suggestion
    logger.info(`Not a GitHub issue...skipping.`);
```

File: src/helpers/comment.ts (hunk 7 above)
Comment (0xcodercrane) [reply src/helpers/comment.ts]: Paragraph is also getting handled in next commits

File: src/helpers/comment.ts (hunk 9 above)
Comment (0xcodercrane) [reply src/helpers/comment.ts:53]: It will be excluded by `itemsToExclude` in next commits

File: src/helpers/user.ts [hunk 15]
```diff
@@ -1,5 +1,4 @@
 import { getBotContext, getLogger } from "../bindings";
-import { Payload } from "../types";
 
 /**
  * @dev Gets the publicly available information about `useranme`
```
Comment (pavlovcik) [inline src/helpers/user.ts]:
```suggestion
 * @dev Gets the publicly available information about `username`
```

**Review by pavlovcik — DISMISSED** [review]

File: src/helpers/user.ts (hunk 15 above)
Comment (0xcodercrane) [reply src/helpers/user.ts]: https://github.com/ubiquity/ubiquibot/pull/305/commits/8229be74f827bbec01adcadf8e1d4829128675d5

**Pull request merged**